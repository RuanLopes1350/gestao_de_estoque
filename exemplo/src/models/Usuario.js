// models/Usuario.js
import mongoose from "mongoose";
import mongoosePaginate from 'mongoose-paginate-v2';

/**
 * Necessario manter os imports para Grupo e Unidade, mesmo sem uso aqui na classe pois o populate do mongoose não funciona se não houver a referencia
 */
import Grupo from './Grupo.js';
import Unidade from './Unidade.js';

class Usuario {
  constructor() {
    const usuarioSchema = new mongoose.Schema(
      {
        nome: { type: String, index: true },
        email: { type: String, unique: true},
        senha: { type: String, select: false },
        link_foto: { type: String },
        ativo: { type: Boolean, default: false },
        tokenUnico: { type: String, select: false }, // Token único para validação de email, recuperação de senha e autenticação
        refreshtoken: { type: String, select: false }, // Refresh token para geração de access token de autenticação longa duração 7 dias para invalidação
        accesstoken: { type: String, select: false }, // Refresh token para  autenticação curta longa 15 minutos para invalidação
        // Referências para Unidades
        codigo_recupera_senha: { type: String, select: false, unique: true }, // Código de recuperação de senha, usado para validar a recuperação de senha do usuário
        exp_codigo_recupera_senha: { type: Date, select: false }, // Data de expiração do código de recuperação de senha, usado para validar a recuperação de senha do usuário

        unidades: [
          {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'unidades' // nome da coleção dentro do banco
          }
        ],

        // Referências para Grupos
        grupos: [
          {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'grupos',
          }
        ],

        /**
         * Permissões personalizadas para cada rota, a rota e o domínio 
         * devem ser únicos e conrrespondentes a rota e domínio do sistema em /rotas
         */
        permissoes: [
          {
            rota: { type: String, index: true, required: true }, // usuários / grupos / unidades / rotas
            dominio: { type: String }, // http://localhost:3000
            ativo: { type: Boolean, default: false },  // false
            buscar: { type: Boolean, default: false },    // false
            enviar: { type: Boolean, default: false },   // false
            substituir: { type: Boolean, default: false },    // false
            modificar: { type: Boolean, default: false },  // false
            excluir: { type: Boolean, default: false }, // false
          }
        ],
      },
      {
        timestamps: true,
        versionKey: false
      }
    );

    // Validação personalizada para garantir que rota + dominio sejam únicos dentro do usuário isso tbm vai ser feito pelo ZOD
    usuarioSchema.pre('save', function (next) {
      const permissoes = this.permissoes;
      const combinacoes = permissoes.map(p => `${p.rota}_${p.dominio}`);
      const setCombinacoes = new Set(combinacoes);

      if (combinacoes.length !== setCombinacoes.size) {
        return next(new Error('Permissões duplicadas encontradas: rota + domínio devem ser únicos dentro de cada usuário.'));
      }

      next();
    });

    usuarioSchema.plugin(mongoosePaginate);

    this.model = mongoose.model('usuarios', usuarioSchema);
  }
}

export default new Usuario().model;
