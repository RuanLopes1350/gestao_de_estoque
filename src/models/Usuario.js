import mongoose from 'mongoose';
import mongoosePaginate from 'mongoose-paginate-v2';

export class Usuario {
    constructor() {
        const usuarioSchema = mongoose.Schema (
            {
                nome_usuario: {type: String, index: true, required: true},
                matricula: {type: String, index: true, required: true, unique: true},
                senha: {type: String, required: true, select: false},
                email: {type: String, required: true, unique: true},
                cargo: {type: String, required: true},
                status: {type: Boolean, default: true},
                acessToken: {type: String, required: true},
                refreshToken: {type: String, required: true},
                data_cadastro: {type: Date, default: Date.now},
                data_atualizacao: {type: Date, default: Date.now},
                data_ultimo_login: {type: Date, default: Date.now},
            },
            {
                versionKey: false,
                timestamps: {createdAt: 'data_cadastro', updatedAt: 'data_atualizacao'},
            }
        )
        usuarioSchema.plugin(mongoosePaginate);
        this.model = mongoose.model('usuarios', usuarioSchema);
    }
}

export default new Usuario().model;