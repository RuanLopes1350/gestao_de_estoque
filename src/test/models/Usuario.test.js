import mongoose from 'mongoose';
import Usuario from '../../models/Usuario';
import { MongoMemoryServer } from 'mongodb-memory-server';
import e from 'express';

let mongoServer;

// Configuração antes de todos os testes
beforeAll(async () => {
    mongoServer = await MongoMemoryServer.create();
    const uri = mongoServer.getUri();

    await mongoose.connect(uri, {
        // Opções de conexão não são necessárias no Mongoose 6+
    });
});

// Limpeza após todos os testes
afterAll(async () => {
    await mongoose.disconnect();
    await mongoServer.stop();
});

// Limpeza após cada teste para garantir isolamento
afterEach(async () => {
    jest.clearAllMocks();
    await Usuario.deleteMany({});
});

describe('Modelo de Usuário', () => {
    it('deve criar um usuário com dados válidos', async () => {
        const userData = {
            nome_usuario: 'Test User',
            matricula: '1234567',
            senha: 'password',
            email: 'testuser@example.com',
            cargo: 'Desenvolvedor',
            status: true,
            accessToken: 'access-token',
            refreshToken: 'refresh-token',
            data_cadastro: new Date(),
            data_atualizacao: new Date(),
            data_ultimo_login: new Date(),
        }

        const user = new Usuario(userData);
        await user.save();

        const savedUser = await Usuario.findById(user._id).select('-senha');
        expect(savedUser.nome_usuario).toBe(userData.nome_usuario);
        expect(savedUser.matricula).toBe(userData.matricula);
        expect(savedUser.senha).toBeUndefined(); // Verifica se a senha não está presente
        expect(savedUser.email).toBe(userData.email);
        expect(savedUser.cargo).toBe(userData.cargo);
        expect(savedUser.status).toBe(userData.status);
        expect(savedUser.accessToken).toBe(userData.accessToken);
        expect(savedUser.refreshToken).toBe(userData.refreshToken);
        expect(savedUser.data_cadastro).toEqual(userData.data_cadastro);
        expect(savedUser.data_atualizacao).toEqual(userData.data_atualizacao);
        expect(savedUser.data_ultimo_login).toEqual(userData.data_ultimo_login);
    });

    it('deve falhar ao criar um usuário sem nome', async () => {
        const userData = {
            matricula: '1234567',
            senha: 'password',
            email: 'testuser@example.com',
            cargo: 'Desenvolvedor',
            status: true,
            accessToken: 'access-token',
            refreshToken: 'refresh-token',
            data_cadastro: new Date(),
            data_atualizacao: new Date(),
            data_ultimo_login: new Date(),
        };

        const user = new Usuario(userData);
        await expect(user.save()).rejects.toThrowErrorMatchingSnapshot();
    });

    it('deve falhar ao criar um usuário sem matrícula', async () => {
        const userData = {
            nome_usuario: 'Test User',
            senha: 'password',
            email: 'testuser@example.com',
            cargo: 'Desenvolvedor',
            status: true,
            accessToken: 'access-token',
            refreshToken: 'refresh-token',
            data_cadastro: new Date(),
            data_atualizacao: new Date(),
            data_ultimo_login: new Date(),
        };

        const user = new Usuario(userData);
        await expect(user.save()).rejects.toThrowErrorMatchingSnapshot();
    })
})