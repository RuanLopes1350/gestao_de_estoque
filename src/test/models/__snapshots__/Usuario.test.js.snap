// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modelo de Usuário deve falhar ao criar um usuário sem matrícula 1`] = `"usuarios validation failed: matricula: Path \`matricula\` is required."`;

exports[`Modelo de Usuário deve falhar ao criar um usuário sem nome 1`] = `"usuarios validation failed: nome_usuario: Path \`nome_usuario\` is required."`;
